File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/external/tax/TaxDsl.java
package com.optivem.shop.dsl.core.usecase.external.tax;

import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.core.shared.UseCaseContext;
import com.optivem.shop.dsl.core.usecase.external.tax.usecases.GetTaxRate;
import com.optivem.shop.dsl.core.usecase.external.tax.usecases.GoToTax;
import com.optivem.shop.dsl.core.usecase.external.tax.usecases.ReturnsTaxRate;
import com.optivem.shop.dsl.driver.port.external.tax.TaxDriver;

public class TaxDsl implements AutoCloseable {
    private final TaxDriver driver;
    private final UseCaseContext context;

    public TaxDsl(TaxDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToTax goToTax() {
        return new GoToTax(driver, context);
    }

    public ReturnsTaxRate returnsTaxRate() {
        return new ReturnsTaxRate(driver, context);
    }

    public GetTaxRate getTaxRate() {
        return new GetTaxRate(driver, context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/tax/TaxDsl.ts
import type { Result } from '../../../../../common/result.js';
import type { TaxDriver } from '../../../../../driver/port/external/tax/tax-driver.js';
import type { TaxErrorResponse } from '../../../../../driver/port/external/tax/dtos/TaxErrorResponse.js';
import type { GetTaxResponse } from '../../../../../driver/port/external/tax/dtos/GetTaxResponse.js';
import type { ReturnsTaxRateRequest } from '../../../../../driver/port/external/tax/dtos/ReturnsTaxRateRequest.js';

export class TaxDsl {
  constructor(private readonly driver: TaxDriver) {}

  async goToTax(): Promise<Result<void, TaxErrorResponse>> {
    return this.driver.goToTax();
  }

  async getTaxRate(country: string): Promise<Result<GetTaxResponse, TaxErrorResponse>> {
    return this.driver.getTaxRate({ country });
  }

  async returnsTaxRate(request: ReturnsTaxRateRequest): Promise<Result<void, TaxErrorResponse>> {
    return this.driver.returnsTaxRate(request);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}
